### PR TITLE
Pickers refactoring

### LINKED ISSUE #5525: Calling triggerTodaySelection throws error
## Description  
Calling `triggerTodaySelection` when date picker is collapsed throws TypeError:
Cannot set property 'viewDate' of undefined

 * igniteui-angular version: 8.1.0
 * browser: all

## Steps to reproduce  

1. Start this [StackBlitz](https://stackblitz.com/edit/date-picker-today-selection-issue?file=src/app/app.component.ts) sample.
2. Click `Today` button.
3. Observe the console

## Result  
ERROR Error: Cannot set property 'viewDate' of undefined
is logged in the console.

## Expected result  
No errors should be logged in the console.

### LINKED ISSUE #6286: min and max properties in the DatePicker component.
## Is your feature request related to a problem? Please describe.
In my case i am using a grid in edit mode.
I need to validate a cell based on value provided in some other cell in the same row.
as there is no min and max properties in date component i cannot validate it efficiently.

## Describe the solution you'd like

If min and max property is present i can just refer the rowData cell  value to the min and max properties and it can be validated using them if any values are entered in some other row cell which is needed for validation

### LINKED ISSUE #6346: DatePicker ARIA
## Is your feature request related to a problem? Please describe.
There are currently no ARIA attributes applied to the date-picker(not in spec either).

## Describe the solution you'd like
The date-picker should implement ARIA attributes per spec (see [this](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/datepicker-dialog)). Note that [the calendar will implement ARI](https://github.com/IgniteUI/igniteui-angular/issues/6272)A on its own.

## Describe alternatives you've considered
N/A

## Additional context
N/A

### LINKED ISSUE #6482: IgxTimePicker refactoring
Dependant on #6271

Related to #7199

[Specification](https://github.com/IgniteUI/igniteui-angular/wiki/Time-Picker)

Once the new `IgxDateTimeEditorDirective` is finished, we should refactor the `time picker` to use the new editor to handle mask parsing, spinning, date validation and any other action related to the `time picker`'s input.

Tasks during refactoring:

- Inheritance of new base parent - introducing new members, renaming old ones (1 day) 
*  Integration of `IgxDateTimeEditorDirective`  (12 days)
     - Code clean up for already existing functionality in the `IgxDatePickerComponent`
     - Refactoring of model binding
     - Move `CalendarContainer` to global date directory
-  Use `DateTimeUtils` 
*  Simplify templating #8665
- New features that will be implemented
    - min and max values (will come from igxDateTimeEditor) #6286 
    - add type property #7196 
    - add ARIA (2 days) #6346, #8932 

    - implementation of Validator interface (1 day)

* Test refactoring (2 days)
*  Bugs that must be resolved (1 day) - #8548
* Migrations for `IgxTimePickerComponent` (2 days)
* Updating of `IgxTimePickerComponent`'s documentation and samples

<hr>
Issues that will be ignored during the refactoring because they could lead to inconsistencies in the product - #6585

### LINKED ISSUE #6483: IgxDatePicker refactoring
Dependant on #6271

Related to #7199

[Specification](https://github.com/IgniteUI/igniteui-angular/wiki/Date-Picker-Specification)

With the new `IgxDateTimeEditorDirective` implemented, we should refactor the `IgxDatePicker` to use the new editor to handle mask parsing, spinning, date validation and any other action related to the date picker's input.

Tasks during refactoring:
- Inheritance of new base parent - introducing new members, renaming old ones (1 day) 
*  Integration of `IgxDateTimeEditorDirective`  (10 days)
     - Code clean up for already existing functionality in the `IgxDatePickerComponent`
     - Refactoring of model binding
     - Move `CalendarContainer` to global date directory
- Have `IgxDateRangePickerComponent` extend `PickersBaseDirective` (2 days)

-  Refactoring of `DatePickerUtils` - will be renamed to `DateTimeUtils` (4 days)
     - Unit tests for `DateTimeUtils`'s public methods
*  Simplify templating #8665
- New features that will be implemented
    - min and max values (will come from igxDateTimeEditor) #6286 
    - add type property #7196 
    - add ARIA (2 days) #6346
    - implementation of Validator interface (1 day)

* Test refactoring (2 days)
* Bugs that must be resolved (1 day) - #8551, #8567, #8475, #8065, #6749, #6925, #5525, #8613, #8913, #8919, #9154, #9189, #9215
* Migrations for `IgxDatePickerComponent` (2 days)
* Updating of `IgxDatePickerComponent`'s documentation and samples

Current estimations are purely for development time.

### LINKED ISSUE #6749: Date picker applies non-date mask when dropdown mode is set after init
## Description
When the `mode` option is changed to `dropdown` after control init, the picker applies a mask that's not a date (the `igxMask` default)

 * igniteui-angular version: 9.0.x
 * browser: N/A

## Steps to reproduce  

1. Open https://stackblitz.com/edit/datepicker-mode-switch
2. Change the mode from the select
3. Focus the input

## Result  
The editor displays the date correctly as "5/15/19", but when you focus there's no value and the placeholder shows the underlying mask as "CCCCCCC..":
image

Also if you clear the editor and focus again, the picker throws a parsing error:

image

## Expected result  
The editor should work the same as if `mode="dropdown"` has been assigned intially, formatting the date as "_5/15/2019" when focused and display an empty mask (`__/__/____`) for cleared.

### LINKED ISSUE #6925: IgxDatePicker allows spinning to negative dates
## Description
captured

 * igniteui-angular version: 8.x, 9.x
 * browser: n/a

## Steps to reproduce  

1. Go to [this sample](https://stackblitz.com/angular/bxoqgbqkpax) or any sample with editable input.
2. Delete part of the year
3. Spin down
,
## Result  
The spinning goes below zero. Additionally, if you try to spin down the month part, you will eventually reach a state where it will keep appending `NaN`.

## Expected result  
I should not be able to spin below zero and there should be no `NaN` values once I spin the month part.

### LINKED ISSUE #6994: Accept ISO 8601 Date-only string as input for IgxDateTimeEditorDirective
## Problem
Calendar and DatePicker use Date class. Date class represents a timestamp which is often troublesome to handle date-only values.
OpenAPI (swagger) defines [format date](https://swagger.io/docs/specification/data-models/data-types/) to be a date-only string (e.g. 2017-07-21).
Currently the Calendar doesn't allow to use information from a Rest endpoint, conforming to the Swagger date format, to display a birthday calendar, or and date picker like `<input type="date" />` allows.
I need to first create Date instances time-shifted to my local time so Calendar is able to display them correctly. The same will happen for other concepts as Holidays, target release date, etc.
e.g. `new Date('2020-03-27')` -> 2020-03-26 21:00:00 GMT-0300 (this makes the Calendar display the wrong date).

## Describe the solution you'd like
Calendar and DatePicker could accept ISO 8601 dates. To internally convert those strings to Date objects or not is an implementation decision, if it's converted it should parse it in a way that it's interpreted as local time, but this should be transparent to the user.
It should also be possible to subscribe to changes using ISO 8601 Date strings, e.g. currently calendar onSelection emits Date objects.

## Describe alternatives you've considered
Create a pipe to transform the date-only strings to local Dates
```
return new Date('2020-03-27' + 'T00:00:00'); // adding time with no timezone makes it parse it as local time.
```
When subscribing to changes I would need to do a conversion to serialize back to date-only strings.
The pipe (input) could be provided by igniteui to use optionally, but this wouldn't solve the subscribe (output) issue.

## Additional context
HTML `<input type="date" />` accepts and emits ISO Date-only strings, just like the swagger format.
By implementing this feature, it allows to switch between `<input type="date">` and `igx-datepicker`, while using the same standard format for values.

Related issues: https://github.com/IgniteUI/igniteui-angular/issues/8256

### LINKED ISSUE #7169: IgxDateTimeEditor should support spin delta
## Is your feature request related to a problem? Please describe.
At the moment, the editor will only spin by `1` which may not cover some user scenarios.

## Describe the solution you'd like
I should be able to provide a number that will determine how the editor will increment or decrement its date parts.

## Describe alternatives you've considered
n/a

## Additional context
n/a

### LINKED ISSUE #7173: Define mechanism to skip not active hours/minutes in igxTimePicker dialog
## Describe the solution you'd like
If igxTimePicker has min and max values, when scrolling into the dialog/dropdown and reaching a max or min value, the disabled hours/minutes/AM or PM should be skipped, and the active time period should become the opposite - min or max value.

## Additional context
Related to #3978

### LINKED ISSUE #7196: Add type property for igxDatePicker, igxTimePicker and igxDateRangePicker
## Is your feature request related to a problem? Please describe.
igxDatePicker, igxTimePicker, and igxDateRangePicker are missing `type` property. Even #6751 is implemented, there is still a high possibility that in addition the `type` property is needed in order to have different types for individual components.

### LINKED ISSUE #7199: Refactor igxDatePicker, igxTimePicker interfaces and define a common one
igxDatePicker implements the following [interface](https://github.com/IgniteUI/igniteui-angular/blob/master/projects/igniteui-angular/src/lib/date-picker/date-picker.common.ts#L5) and igxTimePikcer implements [this](https://github.com/IgniteUI/igniteui-angular/blob/master/projects/igniteui-angular/src/lib/time-picker/time-picker.common.ts#L8) one. After the `IgxDateTimeEditorDirective` is created, these interfaces should be revised. There are properties/methods/events (mode, calendar events) that are common for both, as well for the `IgxRangeDatePickerComponent`, so they should be extracted to a separate interface. Implements IToggleInterface (the latter should be extended and synced within the product)

- [x] Define a common abstract class for the `IgxDatePickerComponent`, `IgxTimePickerComponent` and `IgxDateRangePickerComponent` with base functionality (and possible additional extension for each component)
- [x] Update specification of `IgxDatePickerComponent`
- [ ] Update specification of `IgxTimePickerComponent`

<hr>

- [ ] Refactoring of `IgxDatePickerComponent`
- [ ] Refactoring of `IgxTimePickerComponent` 

`IgxDateTimeEditorDirective` needs to be used as the input for all pickers.

### LINKED ISSUE #8065: Ensure editor works as expected when using a custom template and input field is editable
## Description  
- Currently, when there is a user **template** specified, the editor does not go through the **calculateDate()** method and if the input is not read only, it will cause incorrect component behavior during input.
- Should make sure the input value format is recognized/parsed as a valid one.

Sample - [link](https://stackblitz.com/edit/github-7xy3tu-gc2tkt?file=src%2Fapp%2Fdatepicker-sample-1%2Fdatepicker-sample-1.component.html)

### LINKED ISSUE #8475: DatePicker providing only two digits for the year input is parsed to 19_ _ instead of 20_ _
## Description  
A customer is using a DatePicker in dropdown mode with the default mask. When entering a shorter year such as 5/17/20 it first looks right, but instead of using the current year it uses the year 1920.
 * igniteui-angular version: 10.2.x, 9.1.x
 * browser: any

## Steps to reproduce  

1. Open the following sample: https://stackblitz.com/edit/github-datepikcer-two-digit-year
2. Enter the following value: 06/15/21
3. Click outside the datePicker

## Result  
The created date is 06/15/1921

## Expected result  
The created date should be 06/15/2021

## Attachments  
Attach a sample if available, and screenshots, if applicable.

### LINKED ISSUE #8548: Time picker returns 00:00 when clear the value.
## Description  
Describe the issue.

 * igniteui-angular version: 10.2.1
 * browser: Chrome

## Steps to reproduce  

1. Open the stackblitz, also open the console panel.
https://stackblitz.com/edit/timepicker-null?file=src%2Fapp%2Ftimepicker-sample-1%2Ftimepicker-sample-1.component.ts
2. Set some time to the time picker.
3. Click the close icon to clear the set time.

## Result  
Time picker returns 00:00 as value.

## Expected result  
Time picker returns null the same as the initial value.

### LINKED ISSUE #8551: igx-date-picker breaks its format
## Description  
igx-date-picker breaks its format with a set of keyboard operations

 * igniteui-angular version: 10.2.3
 * browser: n/a

## Steps to reproduce  

1. Run the sample
2. Delete the first digit of the year
3. Hit Down arrow once
4. Hit Right arrow 5 times
5. Hit Down arrow once

## Result  
The format of the value is broken. A string without separator between month and day is presented. e.g. 19/11120
igx-date-picker

## Expected result  
Valid formatted value is presented.

## Attachments  
[igx-date-picker.zip](https://github.com/IgniteUI/igniteui-angular/files/5515939/igx-date-picker.zip)

### LINKED ISSUE #8567: [IgxDatePicker] Day part becomes NaN after clearing the value and hit down arrow key at the day part.
## Description  
[IgxDatePicker] Day part becomes NaN after clearing the value and hit down arrow key at the day part.

 * igniteui-angular version: 10.2.3
 * browser: Chrome

## Steps to reproduce  

1. Open the attached sample in a browser.
2. Click "x" button to clear the value.
3. Click the day part of the mask.
4. Hit down arrow key.

## Result  
"NaN" is displayed in the day part.

## Expected result  
"NaN" is not displayed.
This can be a discussion about to what value the day part should decrement when its value is minimum. It's undefined unless the year and the month are fixed. One option could be "not to wrap and to stop spinning". Another option could be "to wrap to 31 for the time being and to adjust the day value when the editor exits edit mode."

## Attachments  
[cas33657-app1.zip](https://github.com/IgniteUI/igniteui-angular/files/5522838/cas33657-app1.zip)

[cas33657-app1-movie.zip](https://github.com/IgniteUI/igniteui-angular/files/5522848/cas33657-app1-movie.zip)

### LINKED ISSUE #8613: IgxDatePicker throws script error when alphabets are kept pressing in its input text box in IE11
## Description  
IgxDatePicker throws script error when alphabets are kept pressing in its input text box in IE11

 * igniteui-angular version: 10.2.4, 11.0.2
 * browser: IE11

## Steps to reproduce  

1. Start the attached reproducible sample.
2. Click twice the date picker's input box.
3. Keep pressing "a".

## Result  
After starting to keep pressing "a" for a while, "a" is displayed in the text box.
And a script error 'Unable to convert "Invalid Date" into a date' occurs.

## Expected result  
Alphabets are not be accepted.
No script error occurs. 

## Attachments  
[cas33664-app1.zip](https://github.com/IgniteUI/igniteui-angular/files/5564014/cas33664-app1.zip)
cas33664-app1-animation

### LINKED ISSUE #8665: IgxTimePickerComponent, igxHint customization
## Is your feature request related to a problem? 
I want to customize an `<igx-hint>` of `<igx-time-picker>` so I can add custom validation messages, currently it seems the only way to do that is to use an `igxTimePickerTemplate`.

The problem is that `igxTimePickerTemplate` overrides some of your predefined behaviours defined by `InteractionMode` (for my use case I want to keep the numbers validation and the scrolldown wheel for dropdown mode)

## Describe the solution you'd like
Expose the directive `igxHint` so the hint can be passed directly (like `igxLabel` already is)
```
<igx-time-picker format="hh:mm tt" name="end" [mode]="'dropdown'">
  <label igxLabel>End time</label>
  <igx-hint ig class="text-danger" *ngIf="form.controls.end.touched && form.controls.end.hasError('ShouldBeGreater')">
      End time should be greater than start time
  </igx-hint>
</igx-time-picker>
```

## Describe alternatives you've considered
Currently as workaround I'm doing this:
```
<div [class.igx-input-group--invalid]="form.controls.end.touched && form.controls.end.hasError('ShouldBeGreater')">
  <igx-time-picker format="hh:mm tt" name="end" [mode]="'dropdown'">
      <label igxLabel>End time</label>>
  </igx-time-picker>
  <div class="igx-input-group__hint" *ngIf="form.controls.end.touched && form.controls.end.hasError('ShouldBeGreater')">
    <igx-hint class="error-hint">
      End time should be greater than start time
    </igx-hint>
  </div>
</div>
```

### LINKED ISSUE #8913: IgxDatePicker: NaN value on selecting new date when mask and localization are applied
### Description  
When the locale and mask properties of IgxDatePicker are set newly selected date is displayed as NaN

### Steps to reproduce  

1. Open [this](https://stackblitz.com/edit/igx-date-picker-localization-and-formatting?file=src/app/datepicker-sample-4/datepicker-sample-4.component.html) sample
2. Open the igxDatePicker
3. Select some date

### Result  
NaN mask value

### Expected result  
The selected date to be displayed and the mask to be applied correctly
Attach a sample if available, and screenshots, if applicable.

### LINKED ISSUE #8919: igx-date-picker doesn't seems to be working with reactive form
## Question  
I am having trouble with the date picker component when I use it with reactive form, I am setting formControlName and getting the following error

`ERROR TypeError: value.toLocaleDateString is not a function`

date.component.html
`<igx-date-picker [weekStart]="weekStart" [formControlName]="field.protocolDictId" [id]="field.id">
    <label igxLabel>{{field.labelText !== '' ? field.labelText : field.name}}</label>
  </igx-date-picker>`

date.component.ts
`const required = this.field.required ? [Validators.required] : null;
    this.form.addControl(this.field.protocolDictId,
      new FormControl(new Date(Date.now()), {validators: required, updateOn: 'blur'}));`

I have the same code working for other ignite components but having a problem with date-picker component

 * igniteui-angular version: 7.0.1
 * browser: Chrome, Firefox

### LINKED ISSUE #9154: igxDatePicker -  Formatting does not work in dialog mode
## Description  
 Formatting does not work in dialog mode.

 * igniteui-angular version: 
11.1.1

 * browser: 
Google Chorome

## Steps to reproduce  

1. Run the following sample.
https://stackblitz.com/edit/github-2vjjhf?file=src/app/datepicker-sample-3/datepicker-sample-3.component.html
sample

## Result  
Formatting does not work

## Expected result  
Formatting does work

## Attachments  
sample

https://stackblitz.com/edit/github-2vjjhf?file=src/app/datepicker-sample-3/datepicker-sample-3.component.html

### LINKED ISSUE #9189: IgxDatePicker: Date is not getting cleared on clicking clear button when invalid date is entered
### Description  
When trying to clear the date input that is invalid using clear icon, it is not working and instead clear button disappears.

 * igniteui-angular version: 10.2.3
 * browser: Chrome

### Steps to reproduce  

1. Set a disabled dates range.
2. Enter an invalid date (that lies within the disabled date range) using keyboard.
3. Try to clear the input field using clear button.

### Result  
The date that is entered is not getting cleared on clicking clear, instead the clear button itself disappears.

### Expected result  
The date input should get cleared always on clicking the clear button regardless of being valid or invalid.
Also when user tries to enter any date that falls in the disabled dates range, validation should fail.

### Attachments  
Attach a sample if available, and screenshots, if applicable.

### LINKED ISSUE #9215: [IgxDatePicker] When disabled property is changed to true, the label's color becomes red like an invalid state.
## Description  
[IgxDatePicker] When disabled property is changed to true, the label's color becomes red like an invalid state.

 * igniteui-angular version: 11.1.4

## Steps to reproduce  

1. Run the attached sample.
2. Click the input box of the IgxDatePicker.
3. Choose one of the days in the calendar.
4. Turn on the "Disable the date picker".

## Result  
"Birthday" label gets red.

## Expected result  
"Birthday" label gets grayed out.

## Attachments  
[cas34275-app3.zip](https://github.com/IgniteUI/igniteui-angular/files/6219631/cas34275-app3.zip)

### LINKED ISSUE #9377: Exception is thrown in the advanced filtering on editing date condition
## Description  
Introduced with the [Pickers refactoring](#9160)
Exception is thrown in the advanced filtering on editing date condition.

 * igniteui-angular version: 11.2.0-alpha.8
 * browser: NA

## Steps to reproduce  

1. Open the Grid Filtering dev demo.
2. Open the advanced filtering dialog.
3. Create a date filtering condition and commit it.
4. Edit the date condition.

## Result  
Exception is thrown:
ERROR TypeError: Cannot read property 'nativeElement' of undefined
    at enterExpressionEdit (main-es2015.00dec37f962381464056.js:1)

## Expected result  
No exceptions should be thrown.

### PR DESCRIPTION
**Pull request by jackofdiamond5**
Closes #6483 
Closes #6482 
Closes #6994 
Closes #7169 
Closes #7199 
Closes #8665 
Closes #6286
Closes #7196
Closes #6346
Closes #7173

Closes #8551
Closes #8567
Closes  #8475
Closes #8065 
Closes #6749
Closes #6925 
Closes #5525 
Closes #8613 
Closes #8913 
Closes #8919 
Closes #9154
Closes #9189 
Closes #9215
Closes #8548
Closes #9377 

#8256 

### Additional information (check all that apply):
 - [x] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [x] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [x] This PR includes `feature/README.MD` updates for the feature docs
 - [x] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [x] This PR contains breaking changes
 - [x] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [x] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

**Review by valadzhov — DISMISSED** [review]
The incrementendation/decrementation with mouse wheel scroll functionality has to be fixed as discussed. Currently wheel down increments and wheel up decrements and it has to be the other way around.

File: projects/igniteui-angular/src/lib/date-picker/date-picker.component.ts [hunk 1]
```diff
@@ -394,106 +276,31 @@ export class IgxDatePickerComponent implements IDatePicker, ControlValueAccessor
...
-            this._transformedDate = (this._isInEditMode) ? DatePickerUtil.maskToPromptChars(this.inputMask) : '';
-        }
-        return this._transformedDate;
-    }
-
-    /** @hidden @internal */
-    public set transformedDate(value) {
-        this._transformedDate = value;
+    @Input()
+    public get calendarFormat(): IFormattingOptions {
```
Comment (Lipata) [inline projects/igniteui-angular/src/lib/date-picker/date-picker.component.ts:287]: rename back to formatOptions

File: projects/igniteui-angular/src/lib/date-picker/date-picker.component.ts [hunk 2]
```diff
@@ -504,1000 +311,667 @@ export class IgxDatePickerComponent implements IDatePicker, ControlValueAccessor
...
      * ```html
-     * <igx-date-picker mode="dropdown"></igx-date-picker>
+     * <igx-date-picker (onDisabledDate)="onDisabledDate($event)"></igx-date-picker>
      * ```
      */
-    @Input()
-    public mode: InteractionMode = InteractionMode.Dialog;
+    // eslint-disable-next-line @angular-eslint/no-output-on-prefix
+    @Output()
+    @DeprecateProperty('onDisabledDate has been deprecated.')
```
Comment (Lipata) [inline projects/igniteui-angular/src/lib/date-picker/date-picker.component.ts:384]: @DeprecateProperty now working for events. I think can remove it.

File: projects/igniteui-angular/src/lib/date-picker/date-picker.component.ts (hunk 1 above)
Comment (Lipata) [reply projects/igniteui-angular/src/lib/date-picker/date-picker.component.ts:287]: https://github.com/IgniteUI/igniteui-angular/pull/9322/files

File: projects/igniteui-angular/src/lib/date-picker/date-picker.component.html [hunk 3]
```diff
@@ -1,54 +1,29 @@
...
-            igxInput
-            [value]="displayData || ''"
-            [disabled]="disabled"
-            (blur)="onBlur($event)"
-            [tabindex]='editorTabIndex'
-            readonly
-        />
-    </igx-input-group>
-</ng-template>
+    <input class="igx-date-picker__input-date" [displayValuePipe]="this.formatter ? displayValue : null" igxInput
```
Comment (kdinev) [inline projects/igniteui-angular/src/lib/date-picker/date-picker.component.html:6]: Is it intentional that these are no longer in a line per attribute style?

**Review by damyanpetev — APPROVED** [review]

**Review by Lipata — APPROVED** [review]

**Pull request merged**